### PR TITLE
List of deleted forms and ability to restore

### PR DESCRIPTION
**Pull request by ktuite**
There are three main parts of this PR that handle all frontend things related to deleting and restoring forms, including displaying those deleted resources in the audit log (getodk/central#127). 

## 1. Trash section 
This shows a list of the deleted forms under the active form list on the project overview page. This trash section is for admins and project managers, and only shows when there are any deleted forms to show. The forms are in order of when they were deleted (most recent at the bottom) and there is an Undelete button to restore the form (if the form id doesn't conflict with an existing active form id). 

Screen Shot 2021-12-10 at 12 07 07 PM

## 2. Undelete modal
When the user clicks the Undelete button to restore a form, this modal appears, confirming the action of restoring the form and all the form's resources. 

Screen Shot 2021-12-10 at 1 01 53 PM

## 3. Audit log entry names for deleted resources
This part of the PR addresses issue getodk/central#127. It shows names (but no links) for both soft-deleted resources and permanently purged resources. This is based on a change in backend that exposes a `purgedAt` datetime field for purged actees.

Screen Shot 2021-12-10 at 12 45 18 PM

(Still TODO: new tests)

### CODE REVIEW

**Review by ktuite — COMMENTED** [review]
interactive review with @matthew-white

File: src/components/form/trash-list.vue [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+the LICENSE file found in the top-level directory of this distribution and at
+https://www.apache.org/licenses/LICENSE-2.0. No part of ODK Central,
+including this file, may be copied, modified, propagated, or distributed
+except according to the terms contained in the LICENSE file.
+-->
+<template>
+  <div v-if="count > 0">
+    <div id="trash-list-header">
+      <span id="title"><span class="icon-trash"></span>{{ $t('title') }}</span>
+      <span id="trash-count">({{ count }})</span>
```
[the remark this answers is not in the document]
Comment (ktuite) [reply src/components/form/trash-list.vue]: for css purposes, i kept them separate, but i put the parens into a count template. maybe there's a different way to set up the template.

File: test/components/form/restore.spec.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+describe('FormRestore', () => {
+  beforeEach(() => {
+    mockLogin();
+    testData.extendedProjects.createPast(1);
+  });
+
+  it('toggles the modal', () =>
+    load('/projects/1', {}, {
+      deletedForms: () => [
+        new Form({ xmlFormId: '15_days_ago', deletedAt: ago({ days: 15 }).toISO() })
```
Comment (ktuite) [inline test/components/form/restore.spec.js:23]: Don't need to return a form presenter here, just the data of the response

File: test/components/form/restore.spec.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+      };
+      const app = await restore(trashedForms, 111);
+      app.findAllComponents(FormRow).length.should.equal(1);
+      app.findAllComponents(FormTrashRow).length.should.equal(1);
+      app.should.alert('success', 'The Form “Delete Me” has been undeleted.');
+    });
+
+    it('sends the correct requests', async () => {
+      const trashedForms = {
+        111: new Form({ xmlFormId: '111', id: 111, name: 'Delete Me', projectId: 1, deletedAt: ago({ days: 15 }).toISO() })
```
Comment (ktuite) [inline test/components/form/restore.spec.js:130]: don't need these form presenters here, either

File: test/components/form/restore.spec.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+          return { success: true };
+        })
+        .respondWithData(() => Object.values(updatedTrash))
+        .respondWithData(() => testData.extendedForms.sorted());
+    };
+
+    it('shows non-empty active and trash lists updating correctly', async () => {
+      testData.extendedForms.createPast(1);
+      testData.extendedForms.createPast(1);
+      testData.extendedForms.createPast(1);
```
Comment (ktuite) [inline test/components/form/restore.spec.js:88]: createPast(3)

File: test/components/form/trash-list.spec.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        .respondWithData(() => []));
+
+    it('does not request deletedForms if in store already', async () =>
+      mockHttp()
+        .mount(FormTrashList, {
+          requestData: {
+            project: testData.extendedProjects.last(),
+            deletedForms: []
+          }
+        }));
```
Comment (ktuite) [inline test/components/form/trash-list.spec.js:34]: testNoRequest to make this lack of requests explicit

File: test/components/form/trash-list.spec.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import { mockLogin } from '../../util/session';
+
+describe('FormTrashList', () => {
+  beforeEach(() => {
+    mockLogin();
+    testData.extendedProjects.createPast(1);
+  });
+
+  describe('requests', () => {
+    it('requests deletedForms if not yet in store', async () =>
```
Comment (ktuite) [inline test/components/form/trash-list.spec.js:18]: don't need async here or below

File: test/components/form/trash-list.spec.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        new Form({ xmlFormId: '20_days_ago', deletedAt: ago({ days: 20 }).toISO() }),
+        new Form({ xmlFormId: '5_days_ago', deletedAt: ago({ days: 5 }).toISO() })
+      ])
+      .afterResponses(component => {
+        const rows = component.findAllComponents(FormTrashRow);
+        const formXmlIds = rows.wrappers.map(row => row.find('.form-id').text());
+        formXmlIds.should.eql(['20_days_ago', '15_days_ago', '10_days_ago', '5_days_ago']);
+      }));
+
+  it('shows the correct headers', async () =>
```
Comment (ktuite) [inline test/components/form/trash-list.spec.js:85]: don't need async any of these places either

File: test/components/form/trash-row.spec.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+    it('shows the xmlFormId if the form does not have a name', () => {
+      const formData = { xmlFormId: 'f', name: null };
+      mountComponent(formData).get('.form-name').text().should.equal('f');
+    });
+
+    it('shows the deleted timestamp', () => {
+      const formData = { xmlFormId: 'f', name: null, deletedAt: ago({ days: 15 }).toISO() };
+      mountComponent(formData).get('.deleted-date').text().should.match(/^Deleted .+$/);
+      // Should I try to match the date string exactly?
```
Comment (ktuite) [inline test/components/form/trash-row.spec.js:35]: get the DateTime component and check iso prop
or just delete this comment (lower priority)

**Review by matthew-white — APPROVED** [review]

**Review by ktuite — COMMENTED** [review]
interactive review with @matthew-white

**Pull request merged**